### PR TITLE
chore: run license script on pre-commit for staged files

### PR DESCRIPTION
**Pull request by orteth01**
## ✅ Pull Request Checklist:
- [ ] Included link to corresponding [GitHub Issue](https://github.com/gohypergiant/standard-toolkit/issues).
- [ ] The commit message follows [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) [extended](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type) guidelines.
- [ ] Added/updated unit tests and storybook for this change (for bug fixes / features).
- [ ] Added/updated documentation (for bug fixes / features)
- [x] Filled out test instructions.

#### create new file
- create a new file and omit the copyright header
- commit the change
- the copyright header should be added to the staged file and included as part of the commit
- Screenshot 2024-12-19 at 4 52 20 PM

#### modify existing file
- remove the copyright header from an existing file that has it
- commit the change
- the copyright header should be added back to the staged file and included as part of the commit
- Screenshot 2024-12-19 at 4 52 47 PM

#### remove a file
- you should be able to remove a file and commit successfully
- Screenshot 2024-12-19 at 4 51 43 PM

## ❓ Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## 💬 Other information
- after discussion with Brandon, decided to just do the license piece of #4 for now
- as suggested in #4, i used lefthook for the precommit hook
  - see [lefthook documentation](https://evilmartians.github.io/lefthook/intro.html)
- `scripts/license.mjs` deets
    - I adjusted the script to work for files that have an interpreter directive
    - i adjusted the script to accept a '--files' argument, so that we could target only staged files in the pre-commit hook
- we aren't currently adding the header to .yml files. should we?

### CODE REVIEW

**Comment by kalisjoshua** [comment]
I'd rather have a pre-push hook than a pre-commit hook personally. Just attempting to open the conversation on this and  make sure we are all aligned. For me a pre-commit hook is an opportunity to interrupt a devs natural flow (for me it is at least) and I don't like interrupting productivity. This might not be intrusive but if we don't already have pre-commit hooks already, this opens that door.

File: scripts/license.mjs [hunk 1]
```diff
@@ -46,28 +46,46 @@ const COMMENT_STYLES = {
...
+    : ['**/*.{js,ts,tsx,mjs,mdx,md,css}'],
+  {
+    ignore: [
+      '**/node_modules/**',
+      '**/dist/**',
+      '**/README.md',
+      '**/LICENSE.md',
+      '**/CHANGELOG.md',
+      '**/.github/**/*.md',
+      '**/*.yml',
```
[the remark this answers is not in the document]
Comment (brandonjpierce) [reply scripts/license.mjs:70]: I sort of draw the line at: "will this code be included in an application bundle?", if that answer is no we likely don't need a license header

**Comment by belsrc** [comment]
FYSA, since this carried over from a current project. We do (not using lefthook):

```js
  '*.{js,jsx,ts,tsx}': processCodeChunks((chunk) => [
    `prettier ${chunk.join(' ')} --write`,
    `eslint ${chunk.join(' ')} --fix --config "./eslint.config.fast.js"`,
  ]),
  '*.json': processCodeChunks((chunk) => [
    `prettier ${chunk.join(' ')} --write`,
  ]),
  '*.md': ['prettier --write'],
```

on precommit. And 

```json
"lint": "npx turbo run lint",
"test": "npx turbo run test",
"build": "npx turbo run build",
```

on prepush.

I would agree with @kalisjoshua, anything long running I would much prefer to not be on precommit as I commit often and it can get annoying. But also realize the needed for staged files in this case. So a question to you, @orteth01, what does this feel like in practice? Say with ~20 staged files that need the header.

**Comment by belsrc** [comment]
Oh, personally I am fine with those numbers. Current, in project, process is muuuch longer.

**Comment by brandonjpierce** [comment]
> Oh, personally I am fine with those numbers. Current, in project, process is muuuch longer.

@belsrc Yeah that was sort of my push for lefthook in this case when speaking with @orteth01 some of the issues we have on our project can stem from one of two things:

1. Usage of slower tooling e.g. `lint-staged` or `nano-staged`
2. Quantity of files staged (e.g. think about when we commit a large merge conf resolution)

In either scenario you have two parts of the performance equation:

1. How fast is the tooling that is checking and running the scripts against the staged files e.g. `lefthook` or `lint-staged`
2. How fast are the individual tasks within that e.g. `eslint`, `prettier`, `biome`, `oxc` etc etc.

We kind of have a worst case scenario in _all_ scenarios on our current project but in this one we can optimize both sides of that equation technically.

**Comment by brandonjpierce** [comment]
Another point for doing this at the precommit step is that a developer doesn't have to pull the remote in the event we did this work in a CI job and committed then. Less steps involved.

**Comment by kalisjoshua** [comment]
As another perspective - and large just for devils advocate purposes - could we remove it from all source code in the repo and only add it in the published modules are part of that process. I personally dislike seeing it in the code when editing as it takes up screen space for no value and isn't easily collapsed.

Just stirring the pot. Would love to see this but not passionate about it at all. This is the last I hope to mention it.

**Comment by brandonjpierce** [comment]
@kalisjoshua the Apache 2 license specifies we need to include the header in source code:

> Each original source document (code and documentation, but not the LICENSE and NOTICE files) should include [a short license header](https://www.apache.org/legal/src-headers.html#headers) at the top. If the distribution contains documents not covered by an [ICLA](https://www.apache.org/licenses/icla.txt), [CCLA](https://www.apache.org/licenses/cla-corporate.txt) or [Software Grant](https://www.apache.org/licenses/software-grant.txt) (such as third-party libraries), consult the [policy guide](https://www.apache.org/legal/resolved.html).

https://www.apache.org/legal/apply-license.html

**Comment by kalisjoshua** [comment]
Ok. I'm convinced.

The wanna-be-lawyer in me wants to know if the hair could be split that it would be in the released code (published npm modules) while the code used to generate the released code would not need it (non-published github repo).

I know very little about the ramifications of licensing and am completely fine with being overruled by legality.

**Review by kalisjoshua — APPROVED** [review]

**Review by brandonjpierce — APPROVED** [review]

**Pull request merged**